plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java

```diff
@@ -240,55 +253,151 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
-                        .filter(column -> !column.isHidden())
-                        .peek(column -> validateColumnNameForInsert(column.getName()))
-                        .collect(toImmutableList()));
+                handle.getSchemaTableName(),
+                handleColumns,
+                Optional.of(tempTable.getTableName()),
+                Optional.of(pageSinkIdColumn.getName()));
     }
 
     @Override
     public Optional<ConnectorOutputMetadata> finishInsert(ConnectorSession session, ConnectorInsertTableHandle insertHandle, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
     {
+        MongoInsertTableHandle handle = (MongoInsertTableHandle) insertHandle;
+        if (handle.getTemporaryTableName().isPresent()) {
+            finishInsert(handle.getSchemaTableName(), handle.getTemporarySchemaTableName().get(), handle.getPageSinkIdColumnName().get(), fragments);
+        }
+        clearRollback();
         return Optional.empty();
     }
 
+    private void finishInsert(
+            SchemaTableName targetTable,
+            SchemaTableName tempTable,
+            String pageSinkIdColumnName,
+            Collection<Slice> fragments)
+    {
+        Closer closer = Closer.create();
+        closer.register(() -> mongoSession.dropTable(tempTable));
+
+        try {
+            // Create the temporary page sink ID table
+            SchemaTableName pageSinkTable = SchemaTableName.schemaTableName(tempTable.getSchemaName(), generateTemporaryTableName());
+            MongoColumnHandle pageSinkIdColumn = new MongoColumnHandle(pageSinkIdColumnName, TRINO_PAGE_SINK_ID_COLUMN_TYPE, false, Optional.empty());
+            mongoSession.createTable(pageSinkTable, ImmutableList.of(pageSinkIdColumn), Optional.empty());
+            closer.register(() -> mongoSession.dropTable(pageSinkTable));
+
+            // Insert all the page sink IDs into the page sink ID table
+            MongoCollection<Document> pageSinkCollection = mongoSession.getCollection(pageSinkTable);
+            List<Document> pageSinkIds = new ArrayList<>(fragments.size());
+            fragments.forEach(slice -> pageSinkIds.add(new Document(pageSinkIdColumnName, slice.getLong(0))));
+            pageSinkCollection.insertMany(pageSinkIds);
+
+            MongoCollection<Document> tempCollection = mongoSession.getCollection(tempTable);
+            String targetTableName = mongoSession.toRemoteTableName(mongoSession.toRemoteSchemaName(targetTable.getSchemaName()), targetTable.getTableName());
+            tempCollection.aggregate(ImmutableList.of(
+                    lookup(pageSinkTable.getTableName(), pageSinkIdColumnName, pageSinkIdColumnName, "page_sink_id"),
+                    // make this an inner join
+                    match(new BasicDBObject("page_sink_id", new BasicDBObject("$ne", ImmutableList.of()))),
+                    // eliminate page_sink_id field
+                    project(new BasicDBObject("page_sink_id", 0)),
```

Comment:
Could you try `Projections.exclude` method?  Then, we can remove the above comment as the method is descriptive.